src/main/java/com/minecolonies/entity/EntityCitizen.java

```diff
@@ -410,9 +460,19 @@ public void gatherXp()
      */
     public void addExperience(double xp)
     {
-        double maxValue = Integer.MAX_VALUE - citizenData.getExperience();
+        final double citizenHutLevel = getHomeBuilding() == null ? 0 : getHomeBuilding().getBuildingLevel();
+        final double citizenHutMaxLevel = getHomeBuilding() == null ? 1 : getHomeBuilding().getMaxBuildingLevel();
+        if (citizenHutLevel < citizenHutMaxLevel
+              && Math.pow(2.0, citizenHutLevel + 1.0) < this.getExperienceLevel())
+        {
+            return;
+        }
 
+        final double maxValue = Integer.MAX_VALUE - citizenData.getExperience();
         double localXp = xp * skillModifier / EXP_DIVIDER;
+        final double workBuildingLevel = getWorkBuilding() == null ? 0 : getWorkBuilding().getBuildingLevel();
+        final double bonusXp = workBuildingLevel * (1 + citizenHutLevel) / Math.log(this.getExperienceLevel() + 2);
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Cast one of the operands of this addition operation to a "double". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2184)